toolkit/pallets/block-production-log/src/lib.rs

```diff
@@ -0,0 +1,94 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+pub mod mock;
+#[cfg(test)]
+pub mod test;
+
+mod benchmarking;
+
+pub mod weights;
+
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	use sp_block_production_log::*;
+	use sp_consensus_slots::Slot;
+	use sp_runtime::BoundedVec;
+	use sp_std::vec::Vec;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		#[pallet::constant]
+		type MaxLogSize: Get<u32>;
+
+		#[pallet::constant]
+		type MaxProducerIdLength: Get<u32>;
+
+		type BlockProducerId: Member
+			+ Parameter
+			+ MaxEncodedLen
+			+ From<BoundedVec<u8, Self::MaxProducerIdLength>>;
+
+		type WeightInfo: WeightInfo;
+
+		fn current_slot() -> Slot;
+	}
+
+	#[pallet::storage]
+	pub type CurrentBlockProducerId<T: Config> = StorageValue<_, T::BlockProducerId, OptionQuery>;
+
+	#[pallet::storage]
+	pub type Log<T: Config> =
+		StorageValue<_, BoundedVec<(Slot, T::BlockProducerId), T::MaxLogSize>, ValueQuery>;
+
+	#[pallet::inherent]
+	impl<T: Config> ProvideInherent for Pallet<T> {
+		type Call = Call<T>;
+		type Error = InherentError;
+		const INHERENT_IDENTIFIER: InherentIdentifier = INHERENT_IDENTIFIER;
+
+		fn create_inherent(data: &InherentData) -> Option<Self::Call> {
+			let id = data
+				.get_data::<T::BlockProducerId>(&Self::INHERENT_IDENTIFIER)
+				.expect("Block Production Log inherent data not correctly encoded")
+				.expect("Block Production Log inherent data must be provided");
+			Some(Call::append { id })
+		}
+
+		fn is_inherent(call: &Self::Call) -> bool {
+			matches!(call, Call::append { .. })
+		}
+
+		fn is_inherent_required(_: &InherentData) -> Result<Option<Self::Error>, Self::Error> {
+			Ok(Some(Self::Error::InherentRequired))
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight((T::WeightInfo::append(), DispatchClass::Mandatory))]
+		pub fn append(origin: OriginFor<T>, id: T::BlockProducerId) -> DispatchResult {
+			ensure_none(origin)?;
+			let slot = T::current_slot();
+			Log::<T>::try_append((slot, id)).map_err(|_| sp_runtime::DispatchError::BadOrigin)
```

Comment:
TODO: what error when user has not called `take_prefix` on time and the new item doesn't fit in BoundedVec?